File: hamming-error-correction/src/main/java/com/pivovarit/hamming/vavr/HammingDecoder.java
package com.pivovarit.hamming.vavr;

interface HammingDecoder {
    static HammingDecoder vavrDecoder() {
        return new VavrHammingDecoder();
    }

    BinaryString decode(EncodedString input);

    boolean isValid(EncodedString input);
}

File: hamming-error-correction/src/main/kotlin/com/pivovarit/hamming/domain/decode/HammingDecoder.kt
package com.pivovarit.hamming.domain.decode

import com.pivovarit.hamming.domain.BinaryString
import com.pivovarit.hamming.domain.EncodedString
import com.pivovarit.hamming.domain.decode.stateless.SequentialStatelessDecoder

interface HammingDecoder {
    companion object {
        fun sequentialStateless(): HammingDecoder = SequentialStatelessDecoder()
    }

    fun decode(codeWord: EncodedString): BinaryString
    fun isValid(codeWord: EncodedString): Boolean
}